src/System.Private.ServiceModel/src/System/IdentityModel/Claims/X509CertificateClaimSet.cs

```diff
@@ -380,52 +379,108 @@ public IIdentity Identity
-                // Parse: <identifier><delimter><value><separator(s)>
-
-                int delimiterIndex = x509ExtensionFormattedString.IndexOf(subjectName1) - 1;
-                Delimiter = x509ExtensionFormattedString[delimiterIndex];
-
-                // Make an assumption that all characters from the the start of string to the delimiter 
-                // are part of the identifier
-                Identifier = x509ExtensionFormattedString.Substring(0, delimiterIndex);
-
-                int separatorFirstChar = delimiterIndex + subjectName1.Length + 1;
-                int separatorLength = 1;
-                for (int i = separatorFirstChar + 1; i < x509ExtensionFormattedString.Length; i++)
+                try
                 {
-                    // We advance until the first character of the identifier to determine what the
-                    // separator is. This assumes that the identifier assumption above is correct
-                    if (x509ExtensionFormattedString[i] == Identifier[0])
-                    {
-                        break;
-                    }
-                    else
+                    X509Extension x509Extension = new X509Extension(Oid, x509ExtensionBytes, true);
+                    string x509ExtensionFormattedString = x509Extension.Format(false);
+
+                    // Each OS has a different dNSName identifier and delimiter
+                    // On Windows, dNSName == "DNS Name" (localizable), on Linux, dNSName == "DNS"
+                    // e.g.,
+                    // Windows: x509ExtensionFormattedString is: "DNS Name=not-real-subject-name, DNS Name=example.com"
+                    // Linux:   x509ExtensionFormattedString is: "DNS:not-real-subject-name, DNS:example.com"
+                    // Parse: <identifier><delimter><value><separator(s)>
+
+                    int delimiterIndex = x509ExtensionFormattedString.IndexOf(subjectName1) - 1;
+                    _delimiter = x509ExtensionFormattedString[delimiterIndex];
+
+                    // Make an assumption that all characters from the the start of string to the delimiter 
+                    // are part of the identifier
+                    _identifier = x509ExtensionFormattedString.Substring(0, delimiterIndex);
+
+                    int separatorFirstChar = delimiterIndex + subjectName1.Length + 1;
+                    int separatorLength = 1;
+                    for (int i = separatorFirstChar + 1; i < x509ExtensionFormattedString.Length; i++)
                     {
-                        separatorLength++;
+                        // We advance until the first character of the identifier to determine what the
+                        // separator is. This assumes that the identifier assumption above is correct
+                        if (x509ExtensionFormattedString[i] == _identifier[0])
+                        {
+                            break;
+                        }
+                        else
```

Comment:
`else` isn't needed as any following code won't get hit when if clause is true because of the break.